File: scripts/chrome_endpoint.py
import os
from pathlib import Path
from urllib.parse import urlparse, urlunparse

DEFAULT_CHROME_ENDPOINT = "http://127.0.0.1:9222"


def is_wsl() -> bool:
    if os.environ.get("WSL_DISTRO_NAME"):
        return True
    version_path = Path("/proc/version")
    if not version_path.is_file():
        return False
    try:
        return "microsoft" in version_path.read_text(encoding="utf-8").lower()
    except OSError:
        return False


def read_wsl_windows_host() -> str | None:
    resolv_path = Path("/etc/resolv.conf")
    if not resolv_path.is_file():
        return None
    try:
        for line in resolv_path.read_text(encoding="utf-8").splitlines():
            stripped = line.strip()
            if stripped.startswith("nameserver "):
                parts = stripped.split()
                if len(parts) >= 2 and parts[1]:
                    return parts[1]
    except OSError:
        return None
    return None


def rewrite_localhost_for_wsl(endpoint: str) -> str:
    if not is_wsl():
        return endpoint
    parsed = urlparse(endpoint)
    host = (parsed.hostname or "").lower()
    if host not in ("127.0.0.1", "localhost"):
        return endpoint
    win_host = read_wsl_windows_host()
    if not win_host:
        return endpoint
    port = parsed.port
    netloc = f"{win_host}:{port}" if port else win_host
    return urlunparse(parsed._replace(netloc=netloc)).rstrip("/")


def resolve_chrome_endpoint() -> str:
    from_env = os.environ.get("CHROME_DEBUG_ENDPOINT", "").strip()
    raw = from_env if from_env else DEFAULT_CHROME_ENDPOINT
    return rewrite_localhost_for_wsl(raw.rstrip("/"))

File: packages/shared/src/chrome-endpoint.ts
import { readFileSync } from "node:fs";

/** Chrome 远程调试默认地址（Windows 本机） */
export const DEFAULT_CHROME_DEBUG_ENDPOINT = "http://127.0.0.1:9222";


/** 解析 Chrome 调试 URL：环境变量优先；WSL 下将 localhost 改写为 Windows 主机 IP */
export function resolveChromeEndpoint(): string {
  const fromEnv = process.env.CHROME_DEBUG_ENDPOINT?.trim();
  const raw = fromEnv && fromEnv.length > 0
    ? fromEnv.replace(/\/$/, "")
    : DEFAULT_CHROME_DEBUG_ENDPOINT;
  return rewriteLocalhostForWsl(raw);
}


/** Bridge（run.bat / Windows Node）联网搜索用：固定本机 9222，不做 WSL 主机改写 */
export function resolveBridgeChromeEndpoint(): string {
  const fromEnv = process.env.CHROME_DEBUG_ENDPOINT?.trim();
  if (fromEnv && fromEnv.length > 0) {
    return fromEnv.replace(/\/$/, "");
  }
  return DEFAULT_CHROME_DEBUG_ENDPOINT;
}


function isWsl(): boolean {
  if (process.platform !== "linux") {
    return false;
  }
  if (process.env.WSL_DISTRO_NAME) {
    return true;
  }
  try {
    const version = readFileSync("/proc/version", "utf8");
    return /microsoft/i.test(version);
  } catch {
    return false;
  }
}


function readWslWindowsHost(): string | null {
  try {
    const resolv = readFileSync("/etc/resolv.conf", "utf8");
    const match = resolv.match(/^nameserver\s+(\S+)/m);
    return match?.[1] ?? null;
  } catch {
    return null;
  }
}


function rewriteLocalhostForWsl(endpoint: string): string {
  if (!isWsl()) {
    return endpoint;
  }
  try {
    const url = new URL(endpoint);
    const host = url.hostname;
    if (host !== "127.0.0.1" && host !== "localhost") {
      return endpoint;
    }
    const winHost = readWslWindowsHost();
    if (!winHost) {
      return endpoint;
    }
    url.hostname = winHost;
    return url.toString().replace(/\/$/, "");
  } catch {
    return endpoint;
  }
}
